src/app-openid.js

```diff
@@ -0,0 +1,91 @@
+import express from 'express';
+import {
+  errorMiddleware,
+  requestLoggerMiddleware,
+  validateSessionMiddleware,
+} from './util/middlewares.js';
+import { disableOpenID, enableOpenID, isAdmin } from './account-db.js';
+import { loginWithOpenIdFinalize } from './accounts/openid.js';
+import * as UserService from './services/user-service.js';
+
+let app = express();
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+app.use(errorMiddleware);
+app.use(requestLoggerMiddleware);
+export { app as handlers };
+
+app.post('/enable', validateSessionMiddleware, async (req, res) => {
+  if (!isAdmin(res.locals.user_id)) {
+    res.status(403).send({
+      status: 'error',
+      reason: 'forbidden',
+      details: 'permission-not-found',
+    });
+    return;
+  }
+
+  let { error } = (await enableOpenID(req.body)) || {};
+
+  if (error) {
+    res.status(500).send({ status: 'error', reason: error });
+    return;
+  }
+  res.send({ status: 'ok' });
+});
+
+app.post('/disable', validateSessionMiddleware, async (req, res) => {
+  if (!isAdmin(res.locals.user_id)) {
+    res.status(403).send({
+      status: 'error',
+      reason: 'forbidden',
+      details: 'permission-not-found',
+    });
+    return;
+  }
+
+  let { error } = (await disableOpenID(req.body, true)) || {};
+
+  if (error) {
+    res.status(500).send({ status: 'error', reason: error });
+    return;
+  }
+  res.send({ status: 'ok' });
+});
+
+app.get('/config', async (req, res) => {
+  const { cnt: ownerCount } = UserService.getOwnerCount() || {};
+
+  if (ownerCount > 0) {
+    res.status(400).send({ status: 'error', reason: 'already-bootstraped' });
+    return;
+  }
+
+  const auth = UserService.getOpenIDConfig();
+
+  if (!auth) {
+    res
+      .status(500)
+      .send({ status: 'error', reason: 'OpenID configuration not found' });
+    return;
+  }
+
+  try {
+    const openIdConfig = JSON.parse(auth.extra_data);
+    res.send({ openId: openIdConfig });
+  } catch (error) {
+    res
+      .status(500)
+      .send({ status: 'error', reason: 'Invalid OpenID configuration' });
+  }
+});
+
+app.get('/callback', async (req, res) => {
+  let { error, url } = await loginWithOpenIdFinalize(req.query);
```

Comment:
give example of how to implement isValidRedirectUrl, at the table 'auth', when method = 'openid', extra_data has a server_hostname that is the current domain of the application

Reply:
not sure if this will work:
```
if (redirectUrl.hostname === serverHostname) {
```
because the serverhostname must be (protocol)://hostname (example: https://google.com)